contrib/mod_tls.c

```diff
@@ -10472,109 +10407,79 @@ static int tls_verify_cb(int ok, X509_STORE_CTX *ctx) {
   STACK_OF(X509_CRL) *crls = NULL;
-  X509_STORE_CTX *store_ctx = NULL;
-  int n, res;
-
-  /* Unless a revocation store for CRLs was created we cannot do any
-   * CRL-based verification, of course.
-   */
-  if (tls_crl_store == NULL) {
-    return ok;
-  }
-
-  tls_log("%s",
-    "CRL store present, checking client certificate against configured CRLs");
+  int res, verify_error;
 
-  /* Determine certificate ingredients in advance. */
+  verify_error = X509_STORE_CTX_get_error(ctx);
   xs = X509_STORE_CTX_get_current_cert(ctx);
-
   subject = X509_get_subject_name(xs);
-  pr_trace_msg(trace_channel, 15,
-    "verifying cert: subject = '%s'", tls_x509_name_oneline(subject));
-
   issuer = X509_get_issuer_name(xs);
-  pr_trace_msg(trace_channel, 15,
-    "verifying cert: issuer = '%s'", tls_x509_name_oneline(issuer));
-
-  /* OpenSSL provides the general mechanism to deal with CRLs but does not
-   * use them automatically when verifying certificates, so we do it
-   * explicitly here. We will check the CRL for the currently checked
-   * certificate, if there is such a CRL in the store.
-   *
-   * We come through this procedure for each certificate in the certificate
-   * chain, starting with the root-CA's certificate. At each step we've to
-   * both verify the signature on the CRL (to make sure it's a valid CRL)
-   * and its revocation list (to make sure the current certificate isn't
-   * revoked).  But because to check the signature on the CRL we need the
-   * public key of the issuing CA certificate (which was already processed
-   * one round before), we've a little problem. But we can both solve it and
-   * at the same time optimize the processing by using the following
-   * verification scheme (idea and code snippets borrowed from the GLOBUS
-   * project):
-   *
-   * 1. We'll check the signature of a CRL in each step when we find a CRL
-   *    through the _subject_ name of the current certificate. This CRL
-   *    itself will be needed the first time in the next round, of course.
-   *    But we do the signature processing one round before this where the
-   *    public key of the CA is available.
-   *
-   * 2. We'll check the revocation list of a CRL in each step when
-   *    we find a CRL through the _issuer_ name of the current certificate.
-   *    This CRLs signature was then already verified one round before.
-   *
-   * This verification scheme allows a CA to revoke its own certificate as
-   * well, of course.
-   */
 
-  /* Try to retrieve a CRL corresponding to the _subject_ of
-   * the current certificate in order to verify its integrity.
-   */
-  store_ctx = X509_STORE_CTX_new();
-#if OPENSSL_VERSION_NUMBER > 0x000907000L
-  if (X509_STORE_CTX_init(store_ctx, tls_crl_store, NULL, NULL) <= 0) {
-    tls_log("error initializing CRL store context: %s", tls_get_errors());
-    X509_STORE_CTX_free(store_ctx);
-    return ok;
-  }
-#else
-  X509_STORE_CTX_init(store_ctx, tls_crl_store, NULL, NULL);
```

Comment:
Now that we load the configured CRLs into the `SSL_CTX`, we no longer need a separate `X509_STORE_CTX`; we just use the one provided in the callback.  And that means we no longer need the complex implementation from the previous comments.

Less is more.